processing/src/main/java/org/apache/druid/query/filter/TypedInFilter.java

```diff
@@ -0,0 +1,713 @@
+import org.apache.druid.error.InvalidInput;
+import org.apache.druid.java.util.common.ByteBufferUtils;
+import org.apache.druid.java.util.common.IAE;
+import org.apache.druid.java.util.common.StringUtils;
+import org.apache.druid.math.expr.Evals;
+import org.apache.druid.query.cache.CacheKeyBuilder;
+import org.apache.druid.query.filter.vector.VectorValueMatcher;
+import org.apache.druid.query.filter.vector.VectorValueMatcherColumnProcessorFactory;
+import org.apache.druid.segment.ColumnInspector;
+import org.apache.druid.segment.ColumnProcessors;
+import org.apache.druid.segment.ColumnSelectorFactory;
+import org.apache.druid.segment.DimensionHandlerUtils;
+import org.apache.druid.segment.column.ColumnIndexSupplier;
+import org.apache.druid.segment.column.ColumnType;
+import org.apache.druid.segment.column.NullableTypeStrategy;
+import org.apache.druid.segment.column.ValueType;
+import org.apache.druid.segment.filter.Filters;
+import org.apache.druid.segment.index.BitmapColumnIndex;
+import org.apache.druid.segment.index.semantic.Utf8ValueSetIndexes;
+import org.apache.druid.segment.index.semantic.ValueSetIndexes;
+import org.apache.druid.segment.vector.VectorColumnSelectorFactory;
+
+import javax.annotation.Nullable;
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.SortedSet;
+
+public class TypedInFilter extends AbstractOptimizableDimFilter implements Filter
+{
+  private final String column;
+  private final ColumnType matchValueType;
+  @Nullable
+  private final List<?> unsortedValues;
+  private final Supplier<List<?>> lazyMatchValues;
+  @Nullable
+  private final Supplier<SortedSet<ByteBuffer>> lazyMatchValueBytes;
+  @Nullable
+  private final FilterTuning filterTuning;
+  private final Supplier<DruidPredicateFactory> predicateFactorySupplier;
+
+  @JsonIgnore
+  private final Supplier<byte[]> cacheKeySupplier;
+
+  /**
+   * Creates a new filter.
+   *
+   * @param column         column to search
+   * @param values         set of values to match. This collection may be reused to avoid copying a big collection.
+   *                       Therefore, callers should <b>not</b> modify the collection after it is passed to this
+   *                       constructor.
+   * @param matchValueType type of values contained in set
+   * @param filterTuning   optional tuning
+   */
+  @JsonCreator
+  public TypedInFilter(
+      @JsonProperty("column") String column,
+      @JsonProperty("matchValueType") ColumnType matchValueType,
+      @JsonProperty("values") @Nullable List<?> values,
+      @JsonProperty("sortedValues") @Nullable List<?> sortedValues,
+      @JsonProperty("filterTuning") @Nullable FilterTuning filterTuning
+  )
+  {
+    if (NullHandling.replaceWithDefault()) {
+      throw InvalidInput.exception(
```

Comment:
ah yea, that makes sense, will change

Reply:
moved this check to toFilter and optimize methods (the latter since it might rewrite to an invalid equality filter in default value mode if the match set is a single value with the empty string, which is translated into null in default value mode which should use a null filter instead of equality filter)